### PR TITLE
Implement URL support for ActiveMQ Artemis

### PR DESCRIPTION
**Pull request by jbertram**
See gh-10739

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@jbertram Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=24302)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-boot&pullRequestId=24302) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@jbertram Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=24302)!

**Comment by kmandalas** [comment]
@jbertram I see some minor code Formatting violations in CI build, for files `ArtemisConnectionFactoryFactory.java` and `rtemisAutoConfigurationTests.java`. I guess can be resolved by running `gradle format`

**Comment by jbertram** [comment]
@kmandalas the CI build issues have been resolved.

**Comment by kmandalas** [comment]
@pivotal-issuemaster @spring-issuemaster can we proceed with the merge?

**Comment by philwebb** [comment]
@kmandalas We need to wait until we've created a 2.4.x branch before we can merge this.

**Review by snicoll — CHANGES_REQUESTED** [review]
Thanks @jbertram for the proposal. 

While I find the idea of deprecating host/port interesting, it doesn't really address the concerns that we've expressed on the original issue. I've added a few comments, can you please have a look?

**Comment by snicoll** [comment]
I suspect it was flagged with merge amendments as we don't handle deprecated properties this way (don't worry we can take care of it while polishing).

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/artemis/ArtemisConnectionFactoryFactory.java [hunk 1]
```diff
@@ -127,13 +128,21 @@ private boolean isEmbeddedJmsClassPresent() {
...
+		String url = this.properties.getBrokerUrl();
+		if (StringUtils.hasText(url)) {
+			Constructor<T> constructor = factoryClass.getConstructor(String.class);
+			connectionFactory = constructor.newInstance(url);
+		}
+		else {
+			params.put(TransportConstants.HOST_PROP_NAME, this.properties.getHost());
+			params.put(TransportConstants.PORT_PROP_NAME, this.properties.getPort());
+			TransportConfiguration transportConfiguration = new TransportConfiguration(
+					NettyConnectorFactory.class.getName(), params);
```
[the remark this answers is not in the document]
Comment (jbertram) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/artemis/ArtemisConnectionFactoryFactory.java:142]: I thought I had [already addressed](https://github.com/spring-projects/spring-boot/issues/10739#issuecomment-692794114) that concern. To be clear, I don't really see what the concern actually is. The URL and all of its parameters are turned into a `TransportConfiguration` behind the scenes. Everything that can be set directly via a `TransportConfiguration` can be set with URL parameters. In other words there is no "loss of control" that I can see as @wilkinsona mentioned.

This change makes the Artemis integration essentially equivalent to the [ActiveMQ 5.x integration](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/activemq/ActiveMQConnectionFactoryFactory.java#L82) which in my opinion is how it should have been implemented in the first place.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/artemis/ArtemisConnectionFactoryFactory.java [hunk 2]
```diff
@@ -127,13 +128,21 @@ private boolean isEmbeddedJmsClassPresent() {
...
+			Constructor<T> constructor = factoryClass.getConstructor(boolean.class, TransportConfiguration[].class);
+			connectionFactory = constructor.newInstance(false, new TransportConfiguration[] { transportConfiguration });
+		}
 		String user = this.properties.getUser();
```
[the remark this answers is not in the document]
Comment (jbertram) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/artemis/ArtemisConnectionFactoryFactory.java:146]: The user and password cannot be provided via the URL. Therefore those settings are not deprecated. Again, this is how [the ActiveMQ 5.x integration](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/activemq/ActiveMQConnectionFactoryFactory.java#L82) is implemented (i.e. URL, username, & password).

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/artemis/ArtemisConnectionFactoryFactory.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (snicoll) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/artemis/ArtemisConnectionFactoryFactory.java:142]: As far as I can see you did not hence why I am asking here. The code I've referenced makes an explicit setup using `NettyConnectorFactory`. As far as I can see we'd lose that as soon as an url is set.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/artemis/ArtemisConnectionFactoryFactory.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (snicoll) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/artemis/ArtemisConnectionFactoryFactory.java:146]: Sorry, I thought the URL could be used to provide the username and password.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/artemis/ArtemisConnectionFactoryFactory.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jbertram) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/artemis/ArtemisConnectionFactoryFactory.java:142]: I guess I'm not following. What exactly is the concern with losing the explicit `NettyConnectorFactory` setup? The only thing which can be set there is host and port. The URL provides the ability to set host and port as well as any other property.

[the remark this answers is not in the document]
Comment (snicoll) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/artemis/ArtemisConnectionFactoryFactory.java:142]: > What exactly is the concern with losing the explicit NettyConnectorFactory setup?

That's most probably the source of the confusion. When a URL is set, no specific transport is set. When a host and port are set a `NettyConnectorFactory` transport is set. Looking a bit more  `NettyConnectorFactory` seems the default implementation anyway so we'd use that as well. 

That wasn't clear hence why we asked explicitly.

[the remark this answers is not in the document]
Comment (jbertram) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/artemis/ArtemisConnectionFactoryFactory.java:142]: > When a URL is set, no specific transport is set.

That's not accurate. Which "transport" is used is determined by the URL's scheme.

There are only 2 `ConnectorFactory` implementations in Artemis - `NettyConnectorFactory` and `InVMConnectorFactory`. To use the `NettyConnectorFactory` you simply need to use the `tcp://` scheme in the URL. To use the `InVMConnectorFactory` you need to use the `vm://` scheme.

[the remark this answers is not in the document]
Comment (snicoll) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/artemis/ArtemisConnectionFactoryFactory.java:142]: > That's not accurate. 
Sorry, poor choice of words.

I meant that the auto-configuration doesn't do anything special when a broker url is set while it does something explicit (in code) when a host is set. We're very cautious to not introduce any inconsistency and the reason why I asked you here. 

Thanks for the follow-up and the feedback !

**Review by snicoll — APPROVED** [review]

**Comment by snicoll** [comment]
@jbertram thank you for your contribution. I've polished it to prevent users using the (now) deprecated property to get their host/port configuration ignored (which is a smell of you having to add an empty broker-url property [here](https://github.com/spring-projects/spring-boot/pull/24302/commits/b3372414d67820f2fcb455e28cc2ee7028765a5d#diff-548567bbe4e167b990f04940d66c2b533c81191ede75bc3e0caa15eb87e284faR130). We now rather check if no broker url is set by the user and if a host is set we configure as we previously did. Of course, if a broker url is set, we'll always use that and I've added an extra test to validate that scenario.

**Comment by jbertram** [comment]
@snicoll, sounds great! Thanks for the review and help with polishing. Feel free to @ me on anything in the future related to Artemis that you may have questions about.

**Comment by snicoll** [comment]
Will do, thanks again for the great support Justin.

**Pull request closed without merging**